### DISCUSSION TITLE
Help with matcher patterns

### DISCUSSION THREAD

**Discussion by Dan-Essin** [explosion/spaCy#9483] [Help: Coding & Implementations] [answered]
Hello,
Newcomer question. I am working a tutorial based on spacy v2 but my development platform has version 3.1.3 installed.
The tutorial presents this code:
```
# Create a pattern matching two tokens: "iPhone" and "X"
pattern = [{"TEXT": "iPhone"}, {"TEXT": "X"}]
# Add the pattern to the matcher
matcher.add("IPHONE_X_PATTERN", None, pattern)
```
In version 3 that generates errors but this code works:
```
pattern = [[{"TEXT": "iOS"}, {"IS_DIGIT": True}]]
matcher.add("IPHONE_X_PATTERN", pattern)
```
My question is:
Since [{"TEXT": "iPhone"}, {"TEXT": "X"}] is a list of dicts
Why does that fail in version 3 but  [[{"TEXT": "iOS"}, {"IS_DIGIT": True}]], which is a list of a list of dicts, works?

p.s. obviously version 3 has reduced the number of args in matcher.add from 3 to 2.

Thank you

**Comment by kinghuang** [comment, marked as answer]
As you noted, the arguments to [Matcher.add](https://spacy.io/api/matcher#add) changed from spaCy 2 to 3. In spaCy 2, the method signature was:

```python
matcher.add(match_id, on_match, *patterns)
```

Notice the `*` on `*patterns`. This is a [variable-length argument](https://stackabuse.com/variable-length-arguments-in-python-with-args-and-kwargs/). So, after the `match_id` and `on_match` argument, any remaining arguments to `matcher.add(…)` would get collected together as a list for `patterns`.

```python
pattern1 = [{"TEXT": "iPhone", {"TEXT": "X"}]
pattern2 = [{"TEXT": "iOS"}, {"IS_DIGIT": True}]

matcher.add("IPHONE_X_PATTERN", None, pattern1, pattern2)

# The method receives the equivalent of patterns=[pattern1, pattern2],
# from gathering the variable arguments.
```

In spaCy 3, the method signature changed to:

```python
matcher.add(match_id, patterns, on_match=None, greedy=None)
```

Putting aside the differences in the order of the arguments, the lack of a `*` on `patterns` means that it will no longer collect arbitrary arguments passed to the method. This is why you need to put together all the patterns into a list, and pass that to the method.

```python
pattern1 = [{"TEXT": "iPhone", {"TEXT": "X"}]
pattern2 = [{"TEXT": "iOS"}, {"IS_DIGIT": True}]

matcher.add("IPHONE_X_PATTERN", [pattern1, pattern2])

# The method receives patterns=[pattern1, pattern2] like before,
# but because you've passed the list instead of a variable number of arguments being collected.
```
